core/src/common.cpp

```diff
@@ -362,6 +364,435 @@ namespace irods::http
+	auto add_algorithms_to_verifier(
+		jwt::verifier<jwt::default_clock, jwt::traits::nlohmann_json>& _verifier,
+		const jwt::jwks<jwt::traits::nlohmann_json>& _jwks,
+		const jwt::decoded_jwt<jwt::traits::nlohmann_json>& _jwt)
+		-> jwt::verifier<jwt::default_clock, jwt::traits::nlohmann_json>&
+	{
+		namespace logging = irods::http::log;
+
+		const auto alg{_jwt.get_algorithm()};
+
+		// Get the JWK the access token was signed with. This is optional.
+		// See RFC 7515 Section 4.1.4
+		if (_jwt.has_key_id()) {
+			auto key_id{_jwt.get_key_id()};
+			if (_jwks.has_jwk(key_id)) {
+				auto jwk{_jwks.get_jwk(key_id)};
+				add_asymmetric_alg_from_jwk(_verifier, jwk, alg);
+
+				return _verifier;
+			}
+			logging::warn("{}: Could not find the desired [kid] in the JWKs list.", __func__);
+		}
+		// We cannot pick out the specific key used, go through entire list of JWKs
+
+		// The first two characters of 'alg' should give us
+		// enough information to get the algorithm 'family'
+		const auto algorithm_family{alg.substr(0, 2)};
+
+		// 'kty' string to search for in JWK
+		// The only valid values are 'EC', 'RSA', and 'oct'
+		// See (JWA Section 6.1) for the table of valid values
+		std::string search_string;
+
+		// RSA family (JWA Section 3.1)
+		if (algorithm_family == "RS" || algorithm_family == "PS") {
+			// 'kty' string for RSA (JWA Section 6.1)
+			search_string = "RSA";
+		}
+		// EC family (JWA Section 3.1)
+		else if (algorithm_family == "ES") {
+			// 'kty' string for Elliptic Curve (JWA Section 6.1)
+			search_string = "EC";
+		}
+		// Symmetric algo (JWA Section 3.1)
+		else if (algorithm_family == "HS") {
+			add_symmetric_alg(_verifier, alg);
+			return _verifier;
+		}
+
+		// Go through entire key set
+		std::for_each(
+			std::cbegin(_jwks), std::cend(_jwks), [&_verifier, &alg, &search_string](const auto& _jwk) -> void {
+				// Check the optional claims first
+			    // Skip JWK if 'use' is not for signing 'sig'
+			    // See JWK Section 4.2
+				if (_jwk.has_use() && _jwk.get_use() != "sig") {
+					logging::trace("{}: JWK not a signing key, ignoring.", __func__);
+					return;
+				}
+				// JWK might have 'key_ops', try to select keys based off of this
+			    // See JWK Section 4.3
+				else if (_jwk.has_key_operations() && !_jwk.get_key_operations().contains("verify")) {
+					logging::trace("{}: JWK not a key used for verification, ignoring.", __func__);
+					return;
+				}
+
+				if (_jwk.has_algorithm()) {
+					// Add the algorithm if 'alg' matches desired.
+					if (_jwk.get_algorithm() == alg) {
+						add_asymmetric_alg_from_jwk(_verifier, _jwk, alg);
+						return;
+					}
+
+					logging::trace("{}: JWK [alg] does not match JWT [alg], ignoring.", __func__);
+					return;
+				}
+
+				// Fallback to required claim
+				if (_jwk.has_key_type()) {
+					// Extract the 'kty' of the JWK, compare to desired 'kty'
+					if (_jwk.get_key_type() == search_string) {
```

Comment:
It is possible for `search_string` to be empty at this point. I guess it's not really a problem, just wondering if there's any short-circuiting opportunity. :)

Reply:
Failing to match something should be an error higher up. Good catch!

Reply:
Done.